clients/lcdexec/lcdexec.c

```diff
@@ -373,13 +287,145 @@ static int connect_and_setup(void)
 	}
 
 	/* Create our menu */
-	if (menu_sock_send(main_menu, NULL, sock) < 0) {
+	if (main_menu_sock_send(main_menu, sock) < 0) {
 		return -1;
 	}
 
 	return 0;
 }
 
+/**
+ * returns the parameter with the given id, if it is contained in the given command
+ * returns NULL otherwise
+ */
+static CommandParameter * find_parameter(Command * command, int id, CommandParameterType * type)
+{
+	for (kdb_long_long_t i = 0; i < command->parametersSize; i++)
+	{
+		switch (command->parameterTypes[i])
```

Comment:
Can we use the code-generator to avoid mappings like these?

Reply:
What exactly do you mean? How would you solve this differently (without thinking whether or not that is possible through code generation)?

Reply:
> What exactly do you mean?

I only wonder why this mapping is needed. It seems like it was not needed before?

Reply:
It was needed but differently.

Before all menu entries (whether they were menus, commands or parameters) were represented by the same struct. Now we use separate types for different types of menu entries.

Before it was something like:
```c
struct {
    int id;
    union {
       // other stuff
    };
};
```

Now we use something like:
```c
union {
    struct {
       int id;
       // other stuff
    },
   struct {
       int id;
       // other stuff
    };
};
```

Of course this pseudo-code is not correct, but the point is the `id` is inside the union elements, so we need the switch, because it is not guaranteed that the `id` is always stored at the same offset.

The reason for the switch is that the first version would be very difficult to support using code-generation. Even now, we iterate over the menu and add the unique `id`s after Elektra has loaded the whole tree.